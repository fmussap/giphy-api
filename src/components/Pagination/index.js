import React from 'react'
import PropTypes from 'prop-types'
import pagination from '../../utils/pagination'
import Page from './page'

const Pagination = ({ total, activePage, pageLink, onClick, max }) => {
  const apage = activePage === 1 ? activePage : ((activePage - 1) / max) + 1
  const prev = '<'
  const next = '>'
  const nextPage = apage + 1 <= 199 ? apage + 1 : 199
  const prevPage = apage - 1 >= 1 ? apage - 1 : 1
  // the giphy development api key only allows to load 199 pages.
  // this projects is using the same key for production and development
  // for production is necessary to ask for a new type of key.
  const limit = process.env.NODE_ENV === 'development' ? 199 : null

  return (
    <ul className='pagination'>
      <Page className='prev' page={prev} pageLink={pageLink.replace('%page%', prev)} onClick={() => onClick(prevPage)} />
      {pagination({ total: total >= 200 ? (!limit ? total : limit) : total, activePage: apage }).map((page, index) => {
        return (
          <div key={index} className={`pagination-item ${apage === page ? 'active' : ''}`}>
            <Page page={page} pageLink={pageLink.replace('%page%', page)} onClick={onClick} />
          </div>
        )
      })}
      <Page className='next' page={next} pageLink={pageLink.replace('%page%', next)} onClick={() => onClick(nextPage)} />
    </ul>
  )
}

Pagination.defaultProps = {
  pageLink: '',
  activePage: 1
}

Pagination.propTypes = {
  total: PropTypes.number,
  activePage: PropTypes.number,
  pageLink: PropTypes.string,
  onClick: PropTypes.func
}

export default Pagination
