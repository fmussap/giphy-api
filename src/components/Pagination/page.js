import React from 'react'

const Dots = ({ className }) => <div className={className}>...</div>

const Page = ({ page, pageLink, onClick }) => {
  // use '...' when is necessary
  const Component = page === '...' ? Dots : 'a'

  const handleClick = !onClick ? null : (e) => {
    e.preventDefault()
    onClick(page)
  }

  return (
    <Component href={pageLink} onClick={handleClick} className='pagination-link'>
      {page}
    </Component>
  )
}

export default Page
