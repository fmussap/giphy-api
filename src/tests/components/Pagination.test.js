import React from 'react'
import { shallow } from 'enzyme'
import Pagination from '../../components/pagination'

let spy, wrapper
beforeEach(() => {
  spy = jest.fn()
  wrapper = shallow(<Pagination
    onClick={spy}
    total={5}
    activePage={3}
    max={1}
  />)
})

test('should render Pagination correctly', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should handle onClick when prev', () => {
  wrapper.find('Page.prev').simulate('click')
  expect(spy).toHaveBeenLastCalledWith(2)
})

test('should handle onClick when next', () => {
  wrapper.find('Page.next').simulate('click')
  expect(spy).toHaveBeenLastCalledWith(4)
})
