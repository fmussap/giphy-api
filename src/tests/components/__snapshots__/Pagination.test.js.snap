// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Pagination correctly 1`] = `
<ul
  className="pagination"
>
  <Page
    className="prev"
    onClick={[Function]}
    page="<"
    pageLink=""
  />
  <div
    className="pagination-item "
    key="0"
  >
    <Page
      onClick={[Function]}
      page={1}
      pageLink=""
    />
  </div>
  <div
    className="pagination-item "
    key="1"
  >
    <Page
      onClick={[Function]}
      page={2}
      pageLink=""
    />
  </div>
  <div
    className="pagination-item active"
    key="2"
  >
    <Page
      onClick={[Function]}
      page={3}
      pageLink=""
    />
  </div>
  <div
    className="pagination-item "
    key="3"
  >
    <Page
      onClick={[Function]}
      page={4}
      pageLink=""
    />
  </div>
  <div
    className="pagination-item "
    key="4"
  >
    <Page
      onClick={[Function]}
      page={5}
      pageLink=""
    />
  </div>
  <Page
    className="next"
    onClick={[Function]}
    page=">"
    pageLink=""
  />
</ul>
`;
